unified-scheduler-logic/src/lib.rs

```diff
@@ -562,91 +670,305 @@ impl Default for UsageQueueInner {
-                    } else {
-                        count.decrement_self();
+    fn unlock(&mut self, task: &Task, requested_usage: RequestedUsage) -> Option<UsageFromTask> {
+        let mut is_newly_lockable = false;
+        match self {
+            Self::Fifo { current_usage, .. } => {
+                match current_usage {
+                    Some(FifoUsage::Readonly(ref mut count)) => match requested_usage {
+                        RequestedUsage::Readonly => {
+                            if count.is_one() {
+                                is_newly_lockable = true;
+                            } else {
+                                count.decrement_self();
+                            }
+                        }
+                        RequestedUsage::Writable => unreachable!(),
+                    },
+                    Some(FifoUsage::Writable(())) => {
+                        assert_matches!(requested_usage, RequestedUsage::Writable);
+                        is_newly_lockable = true;
                     }
+                    None => unreachable!(),
                 }
-                RequestedUsage::Writable => unreachable!(),
-            },
-            Some(Usage::Writable) => match requested_usage {
-                RequestedUsage::Writable => {
-                    is_unused_now = true;
+                if is_newly_lockable {
+                    *current_usage = None;
                 }
-                RequestedUsage::Readonly => unreachable!(),
-            },
-            None => unreachable!(),
+            }
+            Self::Priority { current_usage, .. } => {
+                match current_usage {
+                    Some(PriorityUsage::Readonly(tasks)) => match requested_usage {
+                        RequestedUsage::Readonly => {
+                            // Don't skip remove()-ing to assert the existence of task
+                            tasks.remove(&task.task_index()).unwrap();
+                            if tasks.is_empty() {
+                                is_newly_lockable = true;
+                            }
+                        }
+                        RequestedUsage::Writable => unreachable!(),
+                    },
+                    Some(PriorityUsage::Writable(_task)) => {
+                        assert_matches!(requested_usage, RequestedUsage::Writable);
+                        is_newly_lockable = true;
+                    }
+                    None => unreachable!(),
+                }
+                if is_newly_lockable {
+                    *current_usage = None;
+                }
+            }
         }
 
-        if is_unused_now {
-            self.current_usage = None;
-            self.blocked_usages_from_tasks.pop_front()
+        if is_newly_lockable {
+            self.pop_usage_from_task()
         } else {
             None
         }
     }
 
     fn push_blocked_usage_from_task(&mut self, usage_from_task: UsageFromTask) {
```

Comment:
i don't think we need the suffix of `_usage_from_task` on all these functions. it's clear from the signature what is being pushed, and we don't push anything else..the distinction is just pushing blocked or not.